### PR TITLE
CCIP-021: Extend Direct Execution Sunset Period

### PR DESCRIPTION
**Pull request by friedger**
This PR
* implements CCIP-021 defined at https://github.com/citycoins/governance/pull/31

### CODE REVIEW

File: contracts/proposals/ccip021-extend-sunset-period-2.clar [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+      ;; MIA vote calculation
+      (avgStacked (/ (+ (scale-up cycle64Amount) (scale-up cycle65Amount)) u2))
+      (scaledVote (/ (* avgStacked MIA_SCALE_FACTOR) MIA_SCALE_BASE))
+    )
+    ;; check that at least one value is positive
+    (asserts! (or (> cycle64Amount u0) (> cycle65Amount u0)) none)
+    ;; return scaled or unscaled value
+    (if scaled (some scaledVote) (some (/ scaledVote VOTE_SCALE_FACTOR)))
+  )
+)
```
Comment (whoabuddy) [inline contracts/proposals/ccip021-extend-sunset-period-2.clar:221]: Update for cycles 80/81

```suggestion
(define-read-only (get-mia-vote (cityId uint) (userId uint) (scaled bool))
  (let
    (
      ;; MAINNET: MIA cycle 80 / first block BTC 834,050 STX 142,301
      ;; cycle 2 / u4500 used in tests
      (cycle80Hash (unwrap! (get-block-hash u4500) none))
      (cycle80Data (at-block cycle80Hash (contract-call? .ccd007-citycoin-stacking get-stacker cityId u2 userId)))
      (cycle80Amount (get stacked cycle80Data))
      ;; MAINNET: MIA cycle 81 / first block BTC 836,150 STX 143,989
      ;; cycle 3 / u6600 used in tests
      (cycle81Hash (unwrap! (get-block-hash u6600) none))
      (cycle81Data (at-block cycle81Hash (contract-call? .ccd007-citycoin-stacking get-stacker cityId u3 userId)))
      (cycle81Amount (get stacked cycle81Data))
      ;; MIA vote calculation
      (avgStacked (/ (+ (scale-up cycle80Amount) (scale-up cycle81Amount)) u2))
      (scaledVote (/ (* avgStacked MIA_SCALE_FACTOR) MIA_SCALE_BASE))
    )
    ;; check that at least one value is positive
    (asserts! (or (> cycle80Amount u0) (> cycle81Amount u0)) none)
    ;; return scaled or unscaled value
    (if scaled (some scaledVote) (some (/ scaledVote VOTE_SCALE_FACTOR)))
  )
)
```

File: contracts/proposals/ccip021-extend-sunset-period-2.clar [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+      (cycle65Amount (get stacked cycle65Data))
+      ;; NYC vote calculation
+      (scaledVote (/ (+ (scale-up cycle64Amount) (scale-up cycle65Amount)) u2))
+    )
+    ;; check that at least one value is positive
+    (asserts! (or (> cycle64Amount u0) (> cycle65Amount u0)) none)
+    ;; return scaled or unscaled value
+    (if scaled (some scaledVote) (some (/ scaledVote VOTE_SCALE_FACTOR)))
+  )
+)
```
Comment (whoabuddy) [inline contracts/proposals/ccip021-extend-sunset-period-2.clar:247]: Update for cycles 80/81

```suggestion
(define-read-only (get-nyc-vote (cityId uint) (userId uint) (scaled bool))
  (let
    (
      ;; NYC cycle 80 / first block BTC 834,050 STX 142,301
      ;; cycle 2 / u4500 used in tests
      (cycle80Hash (unwrap! (get-block-hash u4500) none))
      (cycle80Data (at-block cycle80Hash (contract-call? .ccd007-citycoin-stacking get-stacker cityId u2 userId)))
      (cycle80Amount (get stacked cycle80Data))
      ;; NYC cycle 81 / first block BTC 836,150 STX 143,989
      ;; cycle 3 / u6600 used in tests
      (cycle81Hash (unwrap! (get-block-hash u6600) none))
      (cycle81Data (at-block cycle81Hash (contract-call? .ccd007-citycoin-stacking get-stacker cityId u3 userId)))
      (cycle81Amount (get stacked cycle81Data))
      ;; NYC vote calculation
      (scaledVote (/ (+ (scale-up cycle80Amount) (scale-up cycle81Amount)) u2))
    )
    ;; check that at least one value is positive
    (asserts! (or (> cycle80Amount u0) (> cycle81Amount u0)) none)
    ;; return scaled or unscaled value
    (if scaled (some scaledVote) (some (/ scaledVote VOTE_SCALE_FACTOR)))
  )
)
```

**Review by whoabuddy — CHANGES_REQUESTED** [review]
Found a few things but overall looks great, thank you!

**Comment by whoabuddy** [comment]
Also see the change in #70 to pin the Clarinet version so the tests pass :)

File: contracts/proposals/ccip021-extend-sunset-period-2.clar [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+(define-constant ERR_NO_CITY_ID (err u1706))
+(define-constant ERR_VOTE_FAILED (err u1707))
+
+;; CONSTANTS
+
+(define-constant SELF (as-contract tx-sender))
+(define-constant CCIP_021 {
+  name: "Extend Direct Execute Sunset Period 2",
+  link: "https://github.com/citycoins/governance/blob/feat/ccip-21/ccips/ccip-021/ccip-021-extend-direct-execute-sunset-period-2.md",
+  hash: "", ;; TODO
```
[the remark this answers is not in the document]
Comment (whoabuddy) [reply contracts/proposals/ccip021-extend-sunset-period-2.clar:23]: Actually, it's generally the hash of the commit, e.g. below:
https://github.com/citycoins/governance/pull/31/commits/4e1734ef002105d5b787bb88fedacda4c0dccbf2

Which in the contract we would just use `4e1734ef002105d5b787bb88fedacda4c0dccbf2`.

[the remark this answers is not in the document]
Comment (friedger) [reply contracts/proposals/ccip021-extend-sunset-period-2.clar]: done

**Review by whoabuddy — APPROVED** [review]

**Pull request merged**